pepper-apis/src/main/java/org/broadinstitute/ddp/model/event/ActivityInstanceCreationEventAction.java

```diff
@@ -0,0 +1,102 @@
+    private long studyActivityId;
+
+    public ActivityInstanceCreationEventAction(EventConfiguration eventConfiguration, EventConfigurationDto dto) {
+        super(eventConfiguration, dto);
+        this.studyActivityId = dto.getActivityInstanceCreationStudyActivityId();
+    }
+
+    @Override
+    public void doAction(PexInterpreter pexInterpreter, Handle handle, EventSignal eventSignal) {
+        JdbiActivityInstance jdbiActivityInstance = handle.attach(JdbiActivityInstance.class);
+        JdbiActivityInstanceStatus jdbiActivityInstanceStatus = handle.attach(JdbiActivityInstanceStatus.class);
+        JdbiActivity jdbiActivity = handle.attach(JdbiActivity.class);
+        ActivityStatusEventDao activityStatusEventDao = handle.attach(ActivityStatusEventDao.class);
+        JdbiEventConfigurationOccurrenceCounter jdbiEventConfigurationOccurrenceCounter =
+                handle.attach(JdbiEventConfigurationOccurrenceCounter.class);
+
+        ActivityDto activityDto = jdbiActivity.queryActivityById(studyActivityId);
+
+        // Checking if the maximum number of activities of this type is hit
+        LOG.info("Checking if the maximum number of activities (n = {}) for the study activity (id = {}) is hit",
+                activityDto.getMaxInstancesPerUser(), studyActivityId);
+
+        int numExistingActivities = jdbiActivityInstance.getNumActivitiesForParticipant(
+                studyActivityId,
+                eventSignal.getParticipantId());
+        LOG.info("Found {} existing activity instances for study activity {}", numExistingActivities, studyActivityId);
+
+        if (activityDto.getMaxInstancesPerUser() != null && numExistingActivities >= activityDto.getMaxInstancesPerUser()) {
+            LOG.info(
+                    "The number of existing study activities (n = {}) with id = {} is greater than or equal than"
+                            + " the allowed maximum for this study activity (n = {}), skipping the configuration",
+                    numExistingActivities,
+                    studyActivityId,
+                    activityDto.getMaxInstancesPerUser()
+            );
+            return;
+        }
+
+        // All fine, creating an activity instance
+        String activityInstanceGuid = jdbiActivityInstance.generateUniqueGuid();
+        long newActivityInstanceId = jdbiActivityInstance.insert(
+                studyActivityId,
+                eventSignal.getParticipantId(),
+                activityInstanceGuid,
+                false,
+                Instant.now().toEpochMilli(),
+                null
+        );
+        // Using the low-level facility to avoid infinite recursion
+        jdbiActivityInstanceStatus.insert(
+                newActivityInstanceId,
+                InstanceStatusType.CREATED,
+                Instant.now().toEpochMilli(),
+                eventSignal.getParticipantId()
+        );
+        // queue up any events associated with updating the status to created
```

Comment:
Instead of doing this queuing here, have we considered doing it "recursively"? Perhaps at the end of this method, we can return a new `ActivityStatusChangeEventSignal` (with `CREATED` as target status) to the `EventService`, and then event service will process that new signal.